app/plugin_api_test.go

```diff
@@ -121,7 +125,11 @@ func setupMultiPluginApiTest(t *testing.T, pluginCodes []string, pluginManifests
 }
 
 func setupPluginApiTest(t *testing.T, pluginCode string, pluginManifest string, pluginID string, app *App, c *request.Context) string {
-	return setupMultiPluginApiTest(t, []string{pluginCode}, []string{pluginManifest}, []string{pluginID}, app, c)
+
+	asMain := pluginID != "test_db_driver"
```

Comment:
This whole plugin compilation and testing is where I think we need to improve. Right now, the plugin is compiled as a test binary, and run as a plugin. But there is no way to know the test failures without really inspecting the exit code of the plugin, which is not very easily doable right now.

What can be a good way to do this?

Reply:
Could this be addressed through better logging? Wondering if we could redirect compiled plugin's stderr/stdout to somewhere we can easily access it.

Reply:
Reading the code now I guess that's what you've been doing ehe

Reply:
Yeah, the logging does happen. It's just that currently it's unparseable - `{"level":"debug","ts":1623662916.293494,"caller":"plugin/hclog_adapter.go:52","msg":"plugin process exited","plugin_id":"test_db_driver","wrapped_extras":"path/tmp/667895801/test_db_driver/backend.exepid3738356errorexit status 1"}`. 

A failure would have "errorexit status 1" appended at the end. A success won't. It's quite impossible to automatically detect this.

And moreover, it's also not possible to see the test output. Because you can't pass `-test.v` param without actually modifying the go-plugin code.

Basically all the test does is check against panics - which have actually happened during development btw :P

Reply:
Interestingly, with a little bit of hacking it seems possible. Adding the `-test.v` param to `exec.Command` in `newSupervisor()` is the first (and easy) step, then there's some amount of mangling to do with the way plugin messaging works but if it's for testing purposes it might be okay:

https://user-images.githubusercontent.com/1832946/122107718-abff3480-ce1b-11eb-8998-c7b17e48acd2.mp4

Reply:
Ah nice, I missed the test mode.

```
 // Test, if non-nil, will put plugin serving into "test mode". This is
	// meant to be used as part of `go test` within a plugin's codebase to
	// launch the plugin in-process and output a ReattachConfig.
```

Did you use this?

Reply:
Nope, I wasn't aware of that. It looks like something we might want to try.

I did it in a very hacky way on the plugin side by hijacking stdout/stderr and send any output through `p.API.LogInfo()`.
I say hacky because there's a bit of plumbing needed for the plugin to properly work as you need to intercept the initial message with the rpc address and send that to the original output, otherwise it will timeout waiting for that.

Reply:
Lol okay, I will look into it. Still it doesn't solve the problem of parsing the exit code though. Were you able to find a way to extract that?

Reply:
Not really. I was mainly trying to get some output to help with debugging. I think the best solution to that is to run the server in test mode as you pointed out earlier. Anyway, all this is not a blocker from my side, so feel free to defer it :)